Catanie refactor to remove the duplicate pages for anonymous users

marcoleorato opened the pull request:
## Description

The aim of the PR is to remove all the duplicate functions created specifically for an anonymous users and adapt the already existing one to account for both an anonymous user or a logged in user.

## Motivation

Several part of the catanie code have been duplicated making the site more difficult to maintain as many changes needed to be applied in 2 different pages. There were also some situation in which this was not done leaving, very small, issue that only appeared if you were logged out.
Also the "double" site allow for situations in which a user is logged in but does not realize it since he accessed the anonymous part of the website without knowing.
This also solve issue #807 better than just changing the routing.

## Fixes & Changes:

The flag "searchPublicDataEnabled" is now unused as it's now needed to see public datasets while logged in.
Removed all the anonymous pages and functions.
Modified the layout, header, datasets and datasets details to show the correct elements if an user is not logged in.


<details>
  <summary>Full change list</summary>
  
>#### Removed references to anonymous pages that have been removed in the refactor. ####
>CI/ESS/styles.scss
>CI/MAXIV/maxiv-theme.scss
>CI/RFI/theme.scss
>src/styles.scss

>#### Changed the expected url before login from /anonymous/datasets to /datasets as the anonymous page was removed. ####
>cypress/integration/users-login.spec.js

>#### Removed the anonymous routes, changed the routing a bit to redirect to /datasets instead of /anonymous/datasets. ####
>src/app/app-routing/app-routing.module.ts
>src/app/app-routing/datasets.guard.ts

> #### Added variable to hide part of the UI not visible to anonymous users. Made a few small UI changes to have it work slightly better. Added a reload where needed (mostly visible columns and datasets) when a user login or logout. Added a class "anonymous-site" just below app-app-layout (selector app-app-layout .anonymous-site) that allow for adding extra styles only visible to anonymous user if needed. ####
>src/app/_layout/app-header/app-header.component.html
>src/app/_layout/app-header/app-header.component.scss
>src/app/_layout/app-header/app-header.component.ts
>src/app/_layout/app-layout/app-layout.component.html
>src/app/_layout/app-layout/app-layout.component.scss
>src/app/_layout/app-layout/app-layout.component.ts
>src/app/_layout/app-main-layout/app-main-layout.component.html
>src/app/_layout/app-main-layout/app-main-layout.component.scss
>src/app/_layout/app-main-layout/app-main-layout.component.spec.ts
>src/app/_layout/app-main-layout/app-main-layout.component.ts
>src/app/_layout/layout.module.ts
>src/app/shared/modules/breadcrumb/breadcrumb.component.scss
>src/app/users/login/login.component.html
>src/app/users/login/login.component.scss
>src/app/users/login/login.component.ts
>src/app/datasets/dashboard/dashboard.component.html
>src/app/datasets/dashboard/dashboard.component.scss
>src/app/datasets/dashboard/dashboard.component.ts
>src/app/datasets/dataset-detail/dataset-detail.component.html
>src/app/datasets/dataset-detail/dataset-detail.component.scss
>src/app/datasets/dataset-detail/dataset-detail.component.ts
>src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.html
>src/app/datasets/dataset-details-dashboard/dataset-details-dashboard.component.ts
>src/app/datasets/dataset-table/dataset-table.component.scss
>src/app/datasets/datasets-filter/datasets-filter.component.scss

>#### Removed searchPublicDataEnabled flag as the buttons are now needed to see public data without having to log out. ####
>src/app/datasets/dataset-table-actions/dataset-table-actions.component.html
>src/app/datasets/dataset-table-actions/dataset-table-actions.component.ts

>#### Changed the default value of isPublished in the filters as "" as when lookin at your datasets you see both public and non public datasets you own. ####
>src/app/state-management/actions/datasets.actions.ts
>src/app/state-management/models/index.ts
>src/app/state-management/state/datasets.store.ts

>#### Removed anonymousquery since it's no longer used. ####
>src/app/shared/sdk/services/custom/Dataset.ts
>src/app/shared/sdk/services/custom/DerivedDataset.ts
>src/app/shared/sdk/services/custom/RawDataset.ts

>#### Removed the getAnonymousFilters and readded the isPublished flag to the query. This is now an empty string by default, >in which case it's not sent to the backend as not needed. ####
>src/app/state-management/selectors/datasets.selectors.spec.ts
>src/app/state-management/selectors/datasets.selectors.ts

>#### Files removed as no longer needed. ####
>src/app/_layout/anonymous-layout/anonymous-layout.component.html
>src/app/_layout/anonymous-layout/anonymous-layout.component.scss
>src/app/_layout/anonymous-layout/anonymous-layout.component.spec.ts
>src/app/_layout/anonymous-layout/anonymous-layout.component.ts
>src/app/_layout/login-header/_login-header-theme.scss
>src/app/_layout/login-header/login-header.component.html
>src/app/_layout/login-header/login-header.component.scss
>src/app/_layout/login-header/login-header.component.spec.ts
>src/app/_layout/login-header/login-header.component.ts
>src/app/_layout/login-layout/login-layout.component.html
>src/app/_layout/login-layout/login-layout.component.scss
>src/app/_layout/login-layout/login-layout.component.spec.ts
>src/app/_layout/login-layout/login-layout.component.ts
>src/app/datasets/anonymous-dashboard/anonymous-dashboard.component.html
>src/app/datasets/anonymous-dashboard/anonymous-dashboard.component.scss
>src/app/datasets/anonymous-dashboard/anonymous-dashboard.component.spec.ts
>src/app/datasets/anonymous-dashboard/anonymous-dashboard.component.ts
>src/app/datasets/anonymous-details-dashboard/_anonymous-details-dashboard-theme.scss
>src/app/datasets/anonymous-details-dashboard/anonymous-details-dashboard.component.html
>src/app/datasets/anonymous-details-dashboard/anonymous-details-dashboard.component.scss
>src/app/datasets/anonymous-details-dashboard/anonymous-details-dashboard.component.spec.ts
>src/app/datasets/anonymous-details-dashboard/anonymous-details-dashboard.component.ts
>src/app/datasets/anonymous-details/_anonymous-details-theme.scss
>src/app/datasets/anonymous-details/anonymous-details.component.html
>src/app/datasets/anonymous-details/anonymous-details.component.scss
>src/app/datasets/anonymous-details/anonymous-details.component.spec.ts
>src/app/datasets/anonymous-details/anonymous-details.component.ts

</details>



## Tests included/Docs Updated?

- [ ] Included for each change/fix?
- [x] Passing? (Merge will not be approved unless this is checked) 
- [ ] Docs updated?
- [ ] New packages used/requires npm install? 
- [ ] Toggle added for new features?
- [ ] Requires update of catamel API?

marcoleorato commented:
@nitrosx Friday was a bit late so I'm going to fix it today.
For the e2e tests I already found the possible issue (it should be https://github.com/cypress-io/cypress/issues/2640). The problem is within datasets-public.spec.js at line 30-32, the way the "get" and "contains" is written is not reliable (it does succeed with "npx cypress open" and fail around 50% of times with "npx cypress run") I already fixed it locally removing the "get" and it seems to work. In other test I saw a similar problem that was fixed adding a wait(5000) to be sure the request finished before the get is executed.

I'm waiting to push the changes as I'm also going to add a few unit test to fix the coverage.

nitrosx commented:
@marcoleorato what is the best way to test all the changes that you implemented?
Thanks

marcoleorato commented:
The best way to test would probably to compare the pages with the current develop branch. While logged in with the page at the same route and logged out with the page at the /anonymous/* url.
The changes affect the header, dashboard, dataset details view and the action that happen when you log in/out. But the only "visible" change should be the presence of two additional buttons on the dashboard with a "My Data" and "All Public Data" buttons and not having /anonymous/ in the url. Everything else should be/work in the same way as before.

There are also a few changes to the style that are different from develop, so maybe it could be useful to also have Henrik as reviewer since he was working on the style in the last months.

nitrosx commented:
@marcoleorato: I agree with you. I will ask @henrikjohansson712 to take a look and I will make sure to set aside sometime to test this week.

stephan271 commented:
@marcoleorato could you have a look at the merge conflicts ?

After fixing the merge conflicts: is anything preventing us from merging ? I think this change should enter as soon as possible. Any objections from the colleagues ?

marcoleorato commented:
@stephan271 I have updated the PR branch with the latest changes from develop, now it should be possible to merge without issues if everything else is OK

nguyenlinhlinh commented on src/app/shared/sdk/services/custom/DerivedDataset.ts:
```diff
@@ -1654,37 +1654,6 @@ export class DerivedDatasetApi extends BaseLoopBackApi {
-   * This usually means the response is a `DerivedDataset` object.)
-   * </em>
-   */
-  public anonymousquery(fields: any = {}, limits: any = {}, customHeaders?: Function): Observable<any> {
```
I am worries that removing this has impact on the datasets that returns from the server. For unlogged in user the query will have filter such as isPublished = true. Where does it handle this case ? I just check it quickly compare to our scicat-test and this branch. The result looks different. Also I think this requires changes in the backend

marcoleorato replied:
I'm not completely sure of the question here, the isPublished flag is still added, or not, in datasets.selector.ts based on the value an anonymous user would have an empty string meaning the flag is not added. But having isPublished = true still wouldn't be a mistake, but it would be redundant, as this is the expected way this work, an anonymous user see only published datasets. It would be redundant because, for an anonymous user, the isPublished = true is already enforced on the backend to be sure people don't access datasets they shouldn't be able to access.
Finally there is also the issue of having 2 queries which are almost duplicate of each other which lead to bugs due to the difficulty of keeping both into consideration and moving the changes/fixes on both. For example a fix that was done for the anonymousquery regarding exactly the isPublished flag broke the same flag for the logged in query.

nguyenlinhlinh replied:
Hi, I wrote this comment before I saw the isPublished part later. I am not arguing that we should have two queries that do almost the same but wondering if there are things in the backend that should be updated due to the changes in the frontend.

nguyenlinhlinh commented on src/app/shared/sdk/services/custom/RawDataset.ts:
```diff
@@ -1715,37 +1715,6 @@ export class RawDatasetApi extends BaseLoopBackApi {
-   * This usually means the response is a `RawDataset` object.)
-   * </em>
-   */
-  public anonymousquery(fields: any = {}, limits: any = {}, customHeaders?: Function): Observable<any> {
```
The same comment for this function

nguyenlinhlinh replied:
Just an update from our offline discussion. Me and Marco agreed on that Marco will clean up the unsed anonymous things in catamel and create a PR for that.

nguyenlinhlinh reviewed (changes requested):
Hi @marcoleorato, Thank for the work.

nguyenlinhlinh commented on src/app/state-management/state/datasets.store.ts:
```diff
@@ -57,7 +57,7 @@ export const initialDatasetState: DatasetState = {
     keywords: [],
     scientific: [],
-    isPublished: false
+    isPublished: ""
```
Isn't that better to be true or false instead of string?

marcoleorato replied:
Not for a filter, I had also a talk with Stephan about this. The issue is that isPublished in the Datasets is a flag but in the filter you should be able to choose if it's published (true), it it's not published (false) or if you want both ignoring the tag. A boolean can't fit all three the situation. A "null" would be better, and it's what is normally used, but it can't be used as a value in the material button as it is the default value for "reset", an empty string is the nearest value that can be used as it get removed from the flags sent to the backend from the isNully function in datasets.selector.ts which is what you'd expect.
Also this will, in the future, require a bit of changes in the backend to allow for better filtering but that is a further development and would touch a more delicate part of code.

nguyenlinhlinh replied:
Just an update from our offline discussion. Me and Marco agreed  on that if I find a solution that could use only boolean the Marco will update the code otherwise not.

nguyenlinhlinh replied:
There was another way that could be used to avoid usage of string but it still look bad so no changes needed

nguyenlinhlinh commented on src/app/datasets/dataset-table-actions/dataset-table-actions.component.ts:
```diff
@@ -42,7 +42,7 @@ export class DatasetTableActionsComponent implements OnInit, OnDestroy {
 
   searchPublicDataEnabled = this.appConfig.searchPublicDataEnabled;
-  currentPublicViewMode = false;
+  currentPublicViewMode: boolean | "" = "";
```
This one could requires checking type later and might causing bugs, I would recommend using only boolean.

marcoleorato replied:
Same as the other, you'd need 3 values here.

marcoleorato commented:
A small update regarding this, I am going to make a PR for the backend to remove the anonymousquery from that too. 
Also I found a small bug in how the current user is found so I need to add a small fix to the PR before it's ready to be merged. In short what happen is that if you reload the page the user is found only on the login, datasets and settings page. On the other it looks like you are not logged in. I'll be looking into it as soon as I have some time.

stephan271 commented:
Thanks for taking over the catamel side of it !

marcoleorato commented:
I fixed the small issue I found for the login, also added the change for the issue #836 adding ingestor to the list of functional accounts.

There is also, now, another PR in catamel for the backend part for the removal of anonymousquery.
